### PR TITLE
Update types for fetch-mock v5.8.1

### PR DESCRIPTION
**Pull request by merrywhether**
Please fill in this template.

- [X] Make your PR against the `master` branch.
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [X] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/wheresrhys/fetch-mock>> (no change log in repo sadly)
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `tslint`.

Fixed some lint errors (uses of Object, missing semicolons, and collapsable overloads) and added `once()`, `done()`, `catch()`, and `spy()`.

### CODE REVIEW

File: fetch-mock/index.d.ts [hunk 1]
```diff
@@ -69,7 +69,7 @@ interface MockResponseObject {
 type MockResponse = Response | Promise<Response>
                     | number | Promise<number>
                     | string | Promise<string>
-                    | Object | Promise<Object>
+                    | {} | Promise<{}>
```
Comment (tamird) [inline fetch-mock/index.d.ts]: What's the upshot of this change?

File: fetch-mock/index.d.ts [hunk 2]
```diff
@@ -104,33 +104,41 @@ interface MockOptions {
...
+     */
+    times?: number;
 }
 
 type MockCall = [string, MockRequest];
 
 interface MatchedRoutes {
-    matched: Array<MockCall>;
-    unmatched: Array<MockCall>;
+    matched: MockCall[];
```
Comment (tamird) [inline fetch-mock/index.d.ts]: Again; why is this better?

File: fetch-mock/index.d.ts [hunk 3]
```diff
@@ -140,24 +148,28 @@ interface FetchMockStatic {
...
-    mock(matcher: MockMatcher, response: MockResponse | MockResponseFunction, options: MockOptions): this;
+    mock(options: MockOptions): this;
+
     /**
      * Replaces fetch() with a stub which records its calls, grouped by
        route, and optionally returns a mocked Response object or passes the
-        call through to fetch(). Calls to .mock() can be chained.
-      * @param options The route to mock
+        call through to fetch(). Shorthand for mock() limited to being
+        called one time only. Calls to .mock() can be chained.
```
Comment (tamird) [inline fetch-mock/index.d.ts]: should be "calls to .once()"

File: fetch-mock/index.d.ts [hunk 4]
```diff
@@ -171,6 +183,18 @@ interface FetchMockStatic {
     /**
      * Replaces fetch() with a stub which records its calls, grouped by
        route, and optionally returns a mocked Response object or passes the
+        call through to fetch(). Shorthand for mock() restricted to the GET
+        method and limited to being called one time only. Calls to .mock()
```
Comment (tamird) [inline fetch-mock/index.d.ts]: s/mock/getOnce/

File: fetch-mock/index.d.ts [hunk 5]
```diff
@@ -181,6 +205,18 @@ interface FetchMockStatic {
     /**
      * Replaces fetch() with a stub which records its calls, grouped by
        route, and optionally returns a mocked Response object or passes the
+        call through to fetch(). Shorthand for mock() restricted to the POST
+        method and limited to being called one time only. Calls to .mock()
```
Comment (tamird) [inline fetch-mock/index.d.ts]: s/mock/postOnce/

File: fetch-mock/index.d.ts [hunk 6]
```diff
@@ -191,6 +227,18 @@ interface FetchMockStatic {
     /**
      * Replaces fetch() with a stub which records its calls, grouped by
        route, and optionally returns a mocked Response object or passes the
+        call through to fetch(). Shorthand for mock() restricted to the PUT
+        method and limited to being called one time only. Calls to .mock()
```
Comment (tamird) [inline fetch-mock/index.d.ts]: s/mock/putOnce/

File: fetch-mock/index.d.ts [hunk 7]
```diff
@@ -201,6 +249,18 @@ interface FetchMockStatic {
     /**
      * Replaces fetch() with a stub which records its calls, grouped by
        route, and optionally returns a mocked Response object or passes the
+        call through to fetch(). Shorthand for mock() restricted to the
+        DELETE method and limited to being called one time only. Calls to
+        .mock() can be chained.
```
Comment (tamird) [inline fetch-mock/index.d.ts]: s/mock/deleteOnce/

File: fetch-mock/index.d.ts [hunk 8]
```diff
@@ -209,14 +269,69 @@ interface FetchMockStatic {
       */
     head(matcher: MockMatcher, reponse: MockResponse | MockResponseFunction, options?: MockOptionsMethodHead): this;
     /**
+     * Replaces fetch() with a stub which records its calls, grouped by
+       route, and optionally returns a mocked Response object or passes the
+        call through to fetch(). Shorthand for mock() restricted to the HEAD
+        method and limited to being called one time only. Calls to .mock()
```
Comment (tamird) [inline fetch-mock/index.d.ts]: s/mock/headOnce/

File: fetch-mock/index.d.ts [hunk 9]
```diff
@@ -209,14 +269,69 @@ interface FetchMockStatic {
...
+      * @param response Configures the http response returned by the mock
+      * @param [options] Additional properties defining the route to mock
+      */
+    headOnce(matcher: MockMatcher, reponse: MockResponse | MockResponseFunction, options?: MockOptionsMethodHead): this;
+
+    /**
+     * Replaces fetch() with a stub which records its calls, grouped by
+       route, and optionally returns a mocked Response object or passes the
+        call through to fetch(). Shorthand for mock() restricted to the PATCH
+        method. Calls to .mock() can be chained.
```
Comment (tamird) [inline fetch-mock/index.d.ts]: s/mock/patch/

File: fetch-mock/index.d.ts [hunk 10]
```diff
@@ -209,14 +269,69 @@ interface FetchMockStatic {
...
+      * @param matcher Condition for selecting which requests to mock
+      * @param response Configures the http response returned by the mock
+      * @param [options] Additional properties defining the route to mock
+      */
+    patch(matcher: MockMatcher, reponse: MockResponse | MockResponseFunction, options?: MockOptionsMethodHead): this;
+    /**
+     * Replaces fetch() with a stub which records its calls, grouped by
+       route, and optionally returns a mocked Response object or passes the
+        call through to fetch(). Shorthand for mock() restricted to the PATCH
+        method and limited to being called one time only. Calls to .mock()
```
Comment (tamird) [inline fetch-mock/index.d.ts]: s/mock/patchOnce/

File: fetch-mock/index.d.ts [hunk 11]
```diff
@@ -209,14 +269,69 @@ interface FetchMockStatic {
...
+       response as a normal call to .mock(matcher, response). It can also
+       take an arbitrary function to completely customise behaviour of
+       unmatched calls. If .catch() is called without any parameters then
+       every unmatched call will receive a 200 response.
+      * @param [response] Configures the http response returned by the mock
+      */
+    catch(response?: MockResponse | MockResponseFunction): this;
+
+    /**
+     * Chainable method this records the call history of unmatched calls,
```
Comment (tamird) [inline fetch-mock/index.d.ts]: s/this/that/

File: fetch-mock/index.d.ts [hunk 12]
```diff
@@ -225,50 +340,58 @@ interface FetchMockStatic {
     /**
      * Returns all calls to fetch matching matcherName.
      */
-    calls(matcherName?: string): Array<MockCall>;
-    /**
-     * Returns a Boolean indicating whether fetch was called and a route
-       was matched.
-      */
-    called(): boolean;
+    calls(matcherName?: string): MockCall[];
```
Comment (tamird) [inline fetch-mock/index.d.ts]: again, why `Array<MockCall>`->`MockCall[]`?

**Review by tamird — APPROVED** [review]
Minor comments, but LGTM.

File: fetch-mock/index.d.ts (hunk 1 above)
Comment (merrywhether) [inline fetch-mock/index.d.ts]: This was based on a suggestion from tslint:
`forbidden-types - Avoid using the Object type. Did you mean 'any' or '{}'?`

File: fetch-mock/index.d.ts (hunk 2 above)
Comment (merrywhether) [inline fetch-mock/index.d.ts]: tslint again:
`array-type` - Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.`

File: fetch-mock/index.d.ts (hunk 1 above)
Comment (tamird) [inline fetch-mock/index.d.ts]: Ah, OK.

File: fetch-mock/index.d.ts (hunk 2 above)
Comment (tamird) [inline fetch-mock/index.d.ts]: Ah, OK.

File: fetch-mock/index.d.ts (hunk 12 above)
Comment (merrywhether) [inline fetch-mock/index.d.ts]: tslint again

**Comment by merrywhether** [comment]
Made suggested docstring changes

File: fetch-mock/index.d.ts (hunk 12 above)
Comment (tamird) [inline fetch-mock/index.d.ts]: Ah, OK.

**Comment by tamird** [comment]
👍

**Comment by merrywhether** [comment]
@andy-ms So I tested this in a client-side-only app and it worked there, but there's a bug for a isomorphic apps, which I hadn't tested. I've created a PR to fix it though: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/13878

**Pull request merged**